server/chat-plugins/username-prefixes.ts

```diff
@@ -6,58 +6,79 @@
 import {FS} from '../../lib';
 
 const PREFIXES_FILE = 'config/chat-plugins/username-prefixes.json';
+const WEEK = 24 * 7 * 60 * 60 * 1000;
 
 export class PrefixManager {
+	/** prefix:timeout */
+	timeouts = new Map<ID, NodeJS.Timeout>();
 	constructor() {
 		// after a restart/newly using the plugin, load prefixes from config.js
-		if (!Chat.oldPlugins['username-prefixes']) this.refreshConfig(true);
+		if (!Chat.oldPlugins['username-prefixes']) {
+			this.refreshConfig(true);
+		}
 	}
 
 	save() {
-		FS(PREFIXES_FILE).writeUpdate(() => JSON.stringify(Config.forcedprefixes || {}));
+		FS(PREFIXES_FILE).writeUpdate(() => JSON.stringify(Config.forcedprefixes || []));
 	}
 
 	refreshConfig(configJustLoaded = false) {
-		if (!Config.forcedprefixes) Config.forcedprefixes = {};
+		if (!Config.forcedprefixes) Config.forcedprefixes = [];
 		if (configJustLoaded) {
 			// if we just loaded the config file, ensure that all prefixes are IDs
-			if (Config.forcedprefixes.privacy) Config.forcedprefixes.privacy = Config.forcedprefixes.privacy.map(toID);
-			if (Config.forcedprefixes.modchat) Config.forcedprefixes.modchat = Config.forcedprefixes.modchat.map(toID);
+			if (!Array.isArray(Config.forcedprefixes)) {
+				const convertedPrefixes = [];
+				for (const type in Config.forcedprefixes) {
+					for (const prefix of Config.forcedprefixes[type]) {
+						convertedPrefixes.push({type, prefix, expireAt: Date.now() + WEEK});
+						this.timeouts.set(prefix, setTimeout(() => {
```

Comment:
I think it would be better to keep one master timeout that iterates over all the prefixes and just clears out ones that are dated - would scale better and would be easier to destroy.

Reply:
So... a weekly interval?

Reply:
That's easy I just want to make sure that's what you meant

Reply:
Also, I'm thinking that's not very friendly to the strict nature of prefixes lasting a week; If the server restarts midway through a ladder tour, then the prefix lasts for an additional week and can be used to manipulate game publicity

Reply:
So run it once a day?

Reply:
I would also run the purge on startup on possible. But definitely run the timer at least once a day.

Reply:
One master timeout _is_ better at scale, but there shouldn't be more than 5-10 prefixes at once (and probably more like 2), so I'm not sure that scalability matters much here.

Reply:
I don't think it hurts much to prep for scale. A single timer also is less memory (less to keep track of) and it's easier to manage (don't have to calculate remaining time for each prefix on every startup, etc).